crates/re_query2/src/range/results.rs

```diff
@@ -0,0 +1,207 @@
+///
+/// Use [`RangeResults::get`], [`RangeResults::get_required`] and [`RangeResults::get_optional`]
+/// in order to access the raw results for each individual component.
+#[derive(Debug, Clone)]
+pub struct RangeResults {
+    /// Raw results for each individual component.
+    pub components: IntMap<ComponentName, RangeComponentResults>,
+}
+
+impl Default for RangeResults {
+    #[inline]
+    fn default() -> Self {
+        Self {
+            components: Default::default(),
+        }
+    }
+}
+
+impl RangeResults {
+    #[inline]
+    pub fn contains(&self, component_name: impl Into<ComponentName>) -> bool {
+        self.components.contains_key(&component_name.into())
+    }
+
+    /// Returns the [`RangeComponentResults`] for the specified `component_name`.
+    #[inline]
+    pub fn get(&self, component_name: impl Into<ComponentName>) -> Option<&RangeComponentResults> {
+        self.components.get(&component_name.into())
+    }
+
+    /// Returns the [`RangeComponentResults`] for the specified `component_name`.
+    ///
+    /// Returns an error if the component is not present.
+    #[inline]
+    pub fn get_required(
+        &self,
+        component_name: impl Into<ComponentName>,
+    ) -> crate::Result<&RangeComponentResults> {
+        let component_name = component_name.into();
+        if let Some(component) = self.components.get(&component_name) {
+            Ok(component)
+        } else {
+            Err(DeserializationError::MissingComponent {
+                component: component_name,
+                backtrace: ::backtrace::Backtrace::new_unresolved(),
+            }
+            .into())
+        }
+    }
+
+    /// Returns the [`RangeComponentResults`] for the specified `component_name`.
+    ///
+    /// Returns empty results if the component is not present.
+    #[inline]
+    pub fn get_optional(&self, component_name: impl Into<ComponentName>) -> &RangeComponentResults {
+        let component_name = component_name.into();
+        if let Some(component) = self.components.get(&component_name) {
+            component
+        } else {
+            static DEFAULT: RangeComponentResults = RangeComponentResults::empty();
+            &DEFAULT
+        }
+    }
+}
+
+impl RangeResults {
+    #[doc(hidden)]
+    #[inline]
+    pub fn add(
+        &mut self,
+        component_name: ComponentName,
+        data: impl Iterator<Item = ((TimeInt, RowId), DataCell)>,
+    ) {
+        let (indices, cells): (Vec<_>, Vec<_>) = data
+            .map(|(index, cell)| (index, Promise::new(cell)))
+            .unzip();
+
+        let results = RangeComponentResults { indices, cells };
+        results.sanity_check();
+
+        self.components.insert(component_name, results);
+    }
+}
+
+// ---
+
+/// Uncached results for a particular component when using a range query.
+#[derive(Debug, Clone)]
+pub struct RangeComponentResults {
+    pub indices: Vec<(TimeInt, RowId)>,
+    pub cells: Vec<Promise>,
```

Comment:
Here and everywhere:
```suggestion
    pub promises: Vec<Promise>,